libs/consumption_metrics/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+    /// This is an absolute, per-tenant metric.
+    /// See also prometheus metric RESIDENT_PHYSICAL_SIZE.
+    ResidentSize,
+    /// Size of the remote storage (S3) directory.
+    /// This is an absolute, per-tenant metric.
+    RemoteStorageSize,
+    /// Logical size of the data in the timeline
+    /// This is an absolute, per-timeline metric
+    TimelineLogicalSize,
+    /// Number of bytes proxied
+    /// This is an incremental per endpoint proxy metric
+    /// see also NUM_BYTES_PROXIED_COUNTER
+    ProxyIOBytesPerClient,
+}
+
+impl FromStr for ConsumptionMetricKind {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "written_size" => Ok(Self::WrittenSize),
+            "synthetic_storage_size" => Ok(Self::SyntheticStorageSize),
+            "resident_size" => Ok(Self::ResidentSize),
+            "remote_storage_size" => Ok(Self::RemoteStorageSize),
+            "timeline_logical_size" => Ok(Self::TimelineLogicalSize),
+            "proxy_io_bytes_per_client" => Ok(Self::ProxyIOBytesPerClient),
+            _ => anyhow::bail!("invalid value \"{s}\" for metric type"),
+        }
+    }
+}
+
+impl fmt::Display for ConsumptionMetricKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ConsumptionMetricKind::WrittenSize => "written_size",
+            ConsumptionMetricKind::SyntheticStorageSize => "synthetic_storage_size",
+            ConsumptionMetricKind::ResidentSize => "resident_size",
+            ConsumptionMetricKind::RemoteStorageSize => "remote_storage_size",
+            ConsumptionMetricKind::TimelineLogicalSize => "timeline_logical_size",
+            ConsumptionMetricKind::ProxyIOBytesPerClient => "proxy_io_bytes_per_client",
+        })
+    }
+}
+
+/// Key that uniquely identifies the object, this metric describes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PageserverConsumptionMetricsKey {
+    pub tenant_id: TenantId,
+    pub timeline_id: Option<TimelineId>,
+    pub metric: ConsumptionMetricKind,
+}
+
+pub const CHUNK_SIZE: usize = 1000;
+
+#[derive(serde::Serialize)]
+pub struct EventChunk<'a> {
+    pub events: &'a [PageserverConsumptionMetric],
+}
+
+///
+/// Key that uniquely identifies the object, this metric describes.
+/// Currently, endpoint_id is enough, but this may change later,
+/// so keep it in a named struct.
+///
+/// Both the proxy and the ingestion endpoint will live in the same region (or cell)
+/// so while the project-id is unique across regions the whole pipeline will work correctly
+/// because we enrich the event with project_id in the control-plane endpoint.
+///
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ProxyConsumptionMetricsKey {
+    pub endpoint_id: String,
+    pub direction: String,
+}
+
+#[derive(serde::Serialize)]
+pub struct ProxyEventChunk<'a> {
+    pub events: &'a [ProxyConsumptionMetric],
+}
```

Comment:
I don't think we need this. We wouldn't be able to derive Serialize if there was no way to serialize the slice itself.

Reply:
This is a way to serialize chunks into jsons of the following format
`{"events" : [ {***}, {***} ] }`
instead of manual json constuction.

Reply:
Ah, right, I forgot about this. Still, this doesn't have to be tightly coupled with proxy's metrics:

```rust
#[derive(serde::Serialize)]
struct Events<'a, T> {
    pub events: &'a [T],
}
```